proposals/0000-bidirectional-instances.md

```diff
@@ -0,0 +1,360 @@
+---
+author: Koen Pauwels
+date-accepted: ""
+proposal-number: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/284).
+
+# Bidirectional Type Class Instances
+
+The interaction between type classes and GADTs can lead to some unpleasant
+surprises. The source of this problem is the fact that GHC treats
+instance contexts only as sufficient conditions, while any
+(non-overlapping) instance is also automatically a necessary one.
+
+## Motivation
+
+Consider the following GADT:
+```haskell
+data Term a where
+  Con :: b -> Term b
+  Tup :: Term b -> Term c -> Term (b,c)
+```
+
+A (simplified) implementation for a `Show` class instance might look like this:
+```haskell
+instance Show d => Show (Term d) where
+  show (Con x)   = show x
+  show (Tup x y) = concat ["(", show x, ",", show y, ")"]
+```
+
+Although the implementation seems quite reasonable at first glance, it is
+rejected by the compiler. The wanted constraints from the `Tup` clause
+`(Show b, Show c)` 
+cannot be deduced from the given constraints `(Show d, d ~ (b,c))`.
+This is because GHC considers instance declarations to be unidirectional
+implications from the instance context to the instance head. So the `Show`
+instance for pairs only introduces an axiom of the form
+`∀ a b. (Show a, Show b) => Show (a,b)`.
+But since overlapping instances are
+disallowed by default, instance declarations are in fact bi-implications (the
+conditions in the instance context are not only necessary but also sufficient).
```

Comment:
Not that it's very important, but I think it should be the other way around: the conditions used to be just sufficient and now they are considered necessary, therefore they can be assumed given the head.

Reply:
Thanks!